magento/magento2
# [Performance] Implement command loader to reduce initialization time of Magento CLI

### Description (*)
This change reduces initialization time of the Magento CLI app by allowing deferred initialization of commands until they're needed.

* Implements the Symfony CommandLoaderInterface three times:
  1. `\Magento\Setup\Console\CommandLoader`, a replacement for `Magento_Setup`'s `CommandList`
  1. `\Magento\Framework\Console\CommandLoader`, an alternative means of adding commands outside the `Magento\Setup` namespace
  1. `\Magento\Framework\Console\CommandLoader\Aggregate`, an aggregate command loader, necessary because a Symfony CLI app can only have one Command Loader.

### Related Pull Requests

None

### Fixed Issues (if relevant)

1. Fixes magento/magento2#29266

### Manual testing scenarios (*)

1. Check out the 2.4-develop branch
1. Run a console command through `time`; for exmaple: `time ./bin/magento setup:install --help` and note how long it takes
1. Check out this branch locally (mattwellss:implement-command-loader)
2. Run the same console command again, and note that it is significantly faster than before

### Questions or comments


* Should the "Aggregate" command loader be renamed or implemented a different way? (Extending `\Magento\Framework\ObjectManager\TMap` seems feasible)
* I identified `setup:install` and `setup:config:set` as the two commands causing the most slowdown, so I focused first on migrating the commands in `Magento_setup`. I'm happy to migrate more commands to use the Command Loader, but it seems unnecessary from a performance perspective.
* Should `Magento\Setup\Console\CommandList` be removed entirely, or just have its `getCommandsClasses()` method updated to return an empty array? The latter would preserve BC, in the case that people have extended the class with their own commands.


### Contribution checklist (*)
